br/pkg/lightning/restore/check_info.go

```diff
@@ -658,86 +660,260 @@ func (rc *Controller) SchemaIsValid(ctx context.Context, tableInfo *mydump.MDTab
+	for _, f := range tableMeta.DataFiles {
+		if f.FileMeta.Type != mydump.SourceTypeCSV {
+			continue
+		}
+		_, row, err := rc.readFirstRow(ctx, f.FileMeta)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if len(row) > 0 {
+			rows = append(rows, row)
+		}
+		// only check at most two of all the files
+		if len(rows) >= 2 {
+			break
+		}
+	}
+	if len(rows) == 0 {
+		return nil
+	} else if len(rows) >= 2 {
+		// if the first row in two source files are not the same, they should not be the header line
+		// NOTE: though lightning's logic allows different source files contains different columns or the
+		// order is difference, here we only check if they are exactly the same because this is the common case.
+		if len(rows[0]) != len(rows[1]) {
+			return nil
+		}
+
+		for i := 0; i < len(rows[0]); i++ {
+			if rows[0][i].GetString() != rows[1][i].GetString() {
+				return nil
+			}
+		}
+	}
+
+	// check if some fields are unique and not ignored
+	// if at least one field appears in a unique key, we can sure there is something wrong,
+	// they should be either the header line or the data is duplicated.
+	tableInfo := rc.dbInfos[tableMeta.DB].Tables[tableMeta.Name]
+	tableFields := make(map[string]struct{})
+	uniqueIdxFields := make(map[string]struct{})
+	ignoreColumns, err := rc.cfg.Mydumper.IgnoreColumns.GetIgnoreColumns(tableMeta.DB, tableMeta.Name, rc.cfg.Mydumper.CaseSensitive)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	ignoreColsSet := make(map[string]struct{})
+	for _, col := range ignoreColumns.Columns {
+		ignoreColsSet[col] = struct{}{}
+	}
+	for _, idx := range tableInfo.Core.Indices {
+		if !idx.Unique && !idx.Primary {
+			continue
+		}
+		for _, col := range idx.Columns {
+			if _, ok := ignoreColsSet[col.Name.L]; !ok {
+				uniqueIdxFields[col.Name.L] = struct{}{}
+			}
+		}
+	}
+	for _, f := range tableInfo.Core.Columns {
+		tableFields[f.Name.L] = struct{}{}
+	}
+	if common.TableHasAutoRowID(tableInfo.Core) {
+		tableFields[model.ExtraHandleName.L] = struct{}{}
+	}
+	hasUniqueField := false
+	for _, d := range rows[0] {
+		val := strings.ToLower(d.GetString())
+		if _, ok := tableFields[val]; !ok {
+			return nil
+		}
+		if _, ok := uniqueIdxFields[val]; ok {
+			hasUniqueField = true
+			break
+		}
+	}
+
+	msg := fmt.Sprintf("source csv files contains header row but `mydumper.csv.header` is false, checked table is `%s`.`%s`",
+		tableMeta.DB, tableMeta.Name)
+	level := Warn
+	if hasUniqueField && len(rows) > 1 {
+		level = Critical
+	} else if !checkFieldCompatibility(tableInfo.Core, ignoreColsSet, rows[0]) {
+		// if there are only 1 csv file or there is not unique key, try to check if all columns are compatible with string value
+		level = Critical
+	}
+	rc.checkTemplate.Collect(level, false, msg)
+
+	return nil
+}
+
+func checkFieldCompatibility(tbl *model.TableInfo, ignoreCols map[string]struct{}, values []types.Datum) bool {
+	se := kv.NewSession(&kv.SessionOptions{
+		SQLMode: mysql.ModeStrictTransTables,
```

Comment:
🤔 it should copy the SQL mode from the config, not always using `strict_trans_tables`, because this affects the `CastValue` function below

Reply:
The `ModeStrictTransTables` is by design. the SQL mode in config is likely to be a looser one (like the current default SQL mode of lightning), then failed type cast will be omited.  And when the check logic reaches here, it means the values is the same as the row columns names (likely to be strings in the common case), so the possibility of meeting a false positive case (like a field name is `0000-00-00`?🤔)  is quite low.